lib/pac55xx/ccs.c

```diff
@@ -0,0 +1,271 @@
+}
+void ccs_stclk_sleep_disable(void) {
+	CCSCTL &= ~CCS_CTL_STCLKSLPEN;
+}
+void ccs_set_pclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_PCLKDIV(CCS_CTL_PCLKDIV_MASK)) | CCS_CTL_PCLKDIV(div);
+}
+void ccs_set_aclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_ACLKDIV(CCS_CTL_ACLKDIV_MASK)) | CCS_CTL_ACLKDIV(div);
+}
+void ccs_set_hclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_HCLKDIV(CCS_CTL_HCLKDIV_MASK)) | CCS_CTL_HCLKDIV(div);
+}
+void ccs_pll_enable(void) {
+	CCSPLLCTL |= CCS_PLLCTL_PLLEN;
+}
+void ccs_pll_disable(void) {
+	CCSPLLCTL &= ~CCS_PLLCTL_PLLEN;
+}
+bool ccs_pll_locked(void) {
+	return (CCSPLLCTL & CCS_PLLCTL_PLLLOCK) == CCS_PLLCTL_PLLLOCK;
+}
+void ccs_pll_bypass_enable(void) {
+	CCSPLLCTL |= CCS_PLLCTL_PLLBP;
+}
+void ccs_pll_bypass_disable(void) {
+	CCSPLLCTL &= ~CCS_PLLCTL_PLLBP;
+}
+void ccs_pll_set_outdiv(uint32_t div) {
+	CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLOUTDIV(CCS_PLLCTL_PLLOUTDIV_MASK)) | CCS_PLLCTL_PLLOUTDIV(div);
+}
+void ccs_pll_set_indiv(uint32_t div) {
+	if (div <= 15 && div >= 1) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLINDIV(CCS_PLLCTL_PLLINDIV_MASK)) | CCS_PLLCTL_PLLINDIV(div);
+	}
+}
+void ccs_pll_set_fbdiv(uint32_t div) {
+	if (div <= 16383 && div >= 4) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLFBDIV(CCS_PLLCTL_PLLFBDIV_MASK)) | CCS_PLLCTL_PLLFBDIV(div);
+	}
+}
+void css_pll_config_enable(uint32_t indiv, uint32_t fbdiv, uint32_t outdiv) {
+	ccs_pll_disable();
+	ccs_pll_set_fbdiv(fbdiv);
+	ccs_pll_set_outdiv(outdiv);
+	ccs_pll_set_indiv(indiv);
+	ccs_pll_enable();
+	while (!ccs_pll_locked()) ; /* Wait for PLL lock ~500us */
+}
+uint32_t ccs_get_peripheral_clk_freq(uint32_t periph, uint32_t select) {
+	switch (periph) {
+		case ADC_BASE:
+			return ccs_sclk_frequency;
+		case I2C_BASE:       /* fall through */
+		case USARTA_BASE:    /* fall through */
+		case USARTB_BASE:    /* fall through */
+		case USARTC_BASE:    /* fall through */
+		case USARTD_BASE:    /* fall through */
+		case CAN_BASE:       /* fall through */
+		case GPTIMERA_BASE:  /* fall through */
+		case GPTIMERB_BASE:
+			return ccs_pclk_frequency;
+		case TIMERA_BASE:  /* fall through */
+		case TIMERB_BASE:  /* fall through */
+		case TIMERC_BASE:  /* fall through */
+		case TIMERD_BASE:
+			return (select == 0) ? ccs_pclk_frequency : ccs_aclk_frequency;
+		case MEMCTL_BASE:
+			return (select == 0) ? CCS_ROSC_FREQ : ccs_hclk_frequency;
+		case WWDT_BASE:
+			return (select == 0) ? ccs_frclk_frequency : CCS_ROSC_FREQ;
+		case RTC_BASE:
+			return ccs_frclk_frequency;
+		case CRC_BASE:  /* fall through */
+		case SYS_TICK_BASE:
+			return ccs_hclk_frequency;
+		default:
+			cm3_assert_not_reached();
+			break;
+	}
+	return 0; /* 0 if unsupported */
+}
+
+void ccs_reset_clocks(void) {
+	CCSCTL = CCS_CTL_LDOEN | CCS_CTL_ROSCEN |
+			CCS_CTL_PCLKEN | CCS_CTL_ACLKEN |
+			CCS_CTL_ADCCLKEN | CCS_CTL_STCLKSLPEN;
+	CCSPLLCTL = 0;
+}
+
+int ccs_configure_clocks(const struct ccs_clk_config *config) {
```

Comment:
returning errors is atypical in libopencm clock setup code. Consider adding `cm3_assert` or other mechanism rather than error returns for consistency.

Reply:
removed return value.  errors in config values will now go into `cm3_assert_not_reached()` loop.